### DISCUSSION TITLE
Procedure for fetching data client-side

### DISCUSSION THREAD

**Discussion by SpeedyCraftah** [remix-run/remix#3258] [Q&A] [answered]
This question has probably been asked before but what is the procedure for fetching action data without a form? I have setup an action function to take in the kind of operation to perform so I can have one action function doing multiple things but when fetching data with fetch I am getting a 405 method not allowed.

And so I was wondering if there is a proper procedure for doing so without a form? I have multiple buttons on a page which each make a request to the action function and update the client without refreshing the page.

How I try fetch the data:
```
await fetch(window.location.pathname, { method: "POST", body: new URLSearchParams({ type: "refresh-inbox", ... }).toString(), headers: { "Content-Type": "application/x-www-form-urlencoded" } });
```

Thanks!

**Comment by kiliman** [comment, marked as answer]
```ts
// route.tsx
export let action: ActionFunction = async ({ request }) => {
  const formData = await request.formData()
  const type = formData.get('type')
  switch (type) {
    case 'refresh-inbox': ...
    ...
    default: throw new Response('Bad Request', { status: 400 })
  }  
}

export default function() {

  const fetch = useFetcher()

  const refreshInbox = () => {
    // fetcher will convert object to FormData
    // fetcher will default to submitting to current route if action not specified
    fetcher.submit({ type: 'refresh-inbox' }, { method: 'post', action: 'optional' })
  }
  // you can handle state via fetcher.state and data with fetcher.data
  return <button onClick={refreshInbox}>Refresh Inbox</button>
}

**Reply by kiliman** [reply]
Also in your example, you should not access window inside your components as it will fail on SSR. Use the `useLocation` hook.
```ts
const location = useLocation()
const url = `${location.pathname}${location.search}`
```

**Reply by SpeedyCraftah** [reply]
That makes things a lot simpler than using fetch and takes the burden off of handling state yourself, thanks!
And for the window it is fine as I'm only calling the function inside side effects and events - thanks for the tip!

**Reply by SpeedyCraftah** [reply]
I was also wondering if you know a good way to separate requests by the type? For example a reactive value in the fetcher data.
I need this since I have multiple buttons which submit action data, and as they wait for a response they perform a loading animation and it would look weird for all of the buttons to be doing them when I have only submitted one button.

**Reply by SpeedyCraftah** [reply]
Actually can I do this by instantiating multiple fetchers?

**Reply by kiliman** [reply]
Yes, if you need independent POSTs, then use multiple fetchers. Simplest is to wrap the button in its own component. That way it can define its own fetcher and you can update the UI based on state.

**Reply by kiliman** [reply]
```ts
function FetcherButton({ action, children }: any) {
  const fetcher = useFetcher()
  const onClick = () => {
    fetcher.submit({ action }, { method: 'post' })
  }

  return (
    <button onClick={onClick}>
      {fetcher.state === 'submitting' ? 'Submitting...' : children}
    </button>
  )
}
```
